pkg/operator/ceph/cluster/cleanup.go

```diff
@@ -251,3 +254,118 @@ func (c *ClusterController) getCleanUpDetails(namespace string) (string, string,
 
 	return clusterInfo.MonitorSecret, clusterInfo.FSID, nil
 }
+
+func checkBlockingDependent(typeName, namespace string, instances ...client.Object) (blockDeletion bool) {
+	if len(instances) == 0 {
+		return false
+	}
+	msg := fmt.Sprintf("%s resources exist which should be deleted before the CephCluster in namespace %q:", typeName, namespace)
+	for _, o := range instances {
+		msg = msg + " " + o.GetName()
+	}
+	logger.Errorf(msg)
+	return true
+}
+
+func (c *ClusterController) dependentsBlockDeletion(namespace string) (bool, error) {
+	// TODO: write unit tests
+
+	blockDeletion := false
+	ctx := context.TODO()
+	listOps := metav1.ListOptions{}
+	baseErr := fmt.Sprintf("failed to determine if there are dependent resources for CephCluster in namespace %q", namespace)
+
+	blockingSubresources := []string{
+		// Rook CRs that create Ceph pools
+		"CephBlockPools",
+		"CephFilesystems",
+		"CephObjectStores",
+		// Rook CRs that can use user-created Ceph pools
+		"CephNFSes",
+		"CephClients",
+	}
+
+	// Using a dynamic client is a little more abstract than using the static clients, but we avoid
+	// a lot of copy-pasted code and the human errors that often happen with copy-pasting.
+	dynamicClient, err := dynamic.NewForConfig(c.context.KubeConfig)
+	if err != nil {
+		return true, errors.Wrapf(err, "%s. failed to create new dynamic client for listing subresources", baseErr)
+	}
+
+	for _, sr := range blockingSubresources {
+		lcsr := strings.ToLower(sr)
+		srList, err := dynamicClient.Resource(cephv1.SchemeGroupVersion.WithResource(lcsr)).List(ctx, listOps)
+		if err != nil {
+			return true, errors.Wrapf(err, "%s. failed to list %s resources", baseErr, sr)
+		}
+		if len(srList.Items) > 0 {
+			blockDeletion = true
+			msg := fmt.Sprintf("%s resources exist which should be deleted before the CephCluster in namespace %q:", sr, namespace)
+			for _, inst := range srList.Items {
+				msg = msg + " " + inst.GetName()
+			}
+			logger.Error(msg)
+			// TODO: should we write a status to the CephCluster about this?
+		}
+	}
```

Comment:
You can see what it looks like for copy-pasted code below in comments. CephClient support is not added below. I accidentally made a copy-paste bug when writing this, and I think it's worth the little bit of abstraction to use the dynamic client to avoid this kind of error in the future when we inevitably add another blocking subresource.